### DISCUSSION TITLE
Distribution of messages among offline clients

### DISCUSSION THREAD

**Discussion by mwilniewiec** [vernemq/vernemq#1959] [General]
- VerneMQ Version: 1.12.3
- OS: Ubuntu 18
- Cluster size/standalone: 1

I was testing shared subscriptions and I found one issue that can be problematic for unaware users and I think this could be fixed.
Consider a case where horizontally scaled microservice subscribes to a topic using a shared subscription (for load balancing).
Steps:
1. 5 Instances of a service are subscribed to a shared topic using QOS=1 & clean=false wanting not to lose messages
2. Services get messages distributed among 5 instances **OK**
3. Autoscaler or admin scales microservice to 1
4. Only one online microservice is receiving all the messages now **OK**
5. Microservice dies because of any reason
6. Messages are distributed among 5 offline clients **NOK** & VerneMQ is waiting for those clients to connect
7. Died microservice is stared again and gets 1/5 of messages (published during offline period) **NOK**
8. 4/5 of messages will be published to other subscribers when they connect (if it ever happens) **NOK**

Expected:
In my opinion when all subscribers are offline, qos 1 messages should be treated equally an no client assignment should be done. Client assignment should be done just before sending messages to online clients, not before saving msgs to DB.
With such implementation first client becoming online would receive all messages from the offline period and no msgs would be lost or processed later than expected.

**Comment by ioolkos** [comment]
@mwilniewiec Thank you for your input. It's an often raised point.

While I do not dismiss it as invalid, your picture of the feature is incomplete. Let me, therefore, add a couple of points.

You expect Shared subscriptions to behave as a single queue. This is not the case in VerneMQ. Each subscriber has an exclusive queue it consumes from. This enables the following features:

1. Shared consumers can still control their session settings (like a regular MQTT client). Namely, they should not request message persistence if they do not intend to come back to collect them.
3. Mixed scenarios (some subscribers are persistent, some not)
4. Optimisation in a cluster: we can deliver messages incoming at a specific node to a shared subscriber on that same node. This is called `local` delivery.
5. Above allows for more flexible load balancing policies (called `Shared Subscription Policies`). Currently, the following policies are implemented:  `prefer_local`, `local_only`, `random`, and `prefer_online_to_local`. The last one is new and available in the main branch but not yet part of a release.

So there are multiple criteria in a `Shared Subscription Policy`. Some of those are implicit and admittedly non-intuitive. (`prefer_local` will prefer an offline consumer on the local node to an online consumer at a different node, as an example).
To make things simpler, I have added the `prefer_online_to_local` strategy, which always routes messages to an (any) online consumer in the cluster. It's a more expensive strategy because we need to determine whether consumers are online or offline. But it might be an acceptable trade-off for your use case, as it should allow you to avoid "cold" offline queues.

On a general note: you can use Shared subscribers to do consumer side load-balancing. So you'd use five subscribers when the capacity of your consuming application allows for 1/5 of the message throughput. If one consumer can handle it all, why use Shared Subscribers at all? (rhetorical point only... I get what you want to do)

---
:point_right: Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
:point_right: Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by mwilniewiec** [comment]
I came here with quite good knowledge on how RabbitMQ or Kafka works. Maybe I'm a bit shortsighted when it comes to MQTT but lets take the problem under consideration starting from the very beginning.

What was the real motivation of introducing shared subscriptions in MQTT?
I would say this is **load balancing of messages between multiple customers**.
Why would you ever want to load balance your messages?
1. Thoughtput is so big that vertical scaling of a customer is not enough, you want to scale horizontaly.
2. You want to have high availability of your consumer and when one dies you want others to do his job immediately.
3. Your thoughtput is not that big right now but you are building a horizotally scallable system which is able to handle lots of users/iot devices and get ready for the future.

Additionally new trends like microservices, kubernetes/docker swarm, autoscaling makes it very simple to scale up your services to have multiple consumers.
I would say that vast majority of shared subscription clients are horizontally scaled services (where every each of it is the same, in rare cases twin-similar). Its designed for a many publishers, one subscriber problem.

What was the real motivation of doing clustering in VerneMQ?
I would say 3 points from above are also valid. 
I also think that broker cluster should give you those benefits but should be totally transparent when it comes to functionality and how clients use the broker. When you a are a publisher or a subscriber you don't care if broker is a single instance or its a cluster of 10.

I can understand that current implementation works the way you described but its a bit hard for me to find a business case where multiple consumers sharing one subscription are set in a mixed mode where some are persistent, some are not. 
A consumer using a shared subscription will get only a random part of the whole stream. When he is really unlucky, he might get almost no messages.
I can't also find a business case where online client is doing nothing and messages are sent to a local offline client queue and waiting for him to connect (if it ever happens).
Of course in case clients configure such a monster - broker must behave somehow but I would say its a minor problem how it's solved.

My thoughts:
The tool/feature/program should mainly solve the problem it is created for. Implementation under the hood is just a detail.
If only motivation of `Shared Subscription Policies` was performance optimization I would say its unnecessary complexity and it breaks the rule of standalone broker working the same as cluster of brokers.

From my perspective the basic rule for Shared subscriptions should be `Distribute between online clients`.
With optional settings like: `random_local_first`, `random`, `round_robin`, `idle_first`.
I will only add that this basic rule is valid when you look at other brokers like RMQ or Kafka.

**Reply by ioolkos** [reply]
@mwilniewiec thanks for your input.
Just as a quick reference for now: https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901250
As a server implementor, the above is what you have to implement.

**Reply by mwilniewiec** [reply]
I checked the docs and I couldn't find too many details related to our discussion.
In that case its hard to say who is right :)

> Shared Subscriptions are defined at the scope of the MQTT Server, rather than of a Session.

Maybe its this sentence is not too detailed but I think that VerneMQ should in general threat shared subscriptions more globally rather then session-related. Clustering is a totally different problem which is not a scope of MQTT so as I said before... Cluster of brokers should work exactly the same as a single broker when it comes to talking with MQTT clients.

>   If the Server is in the process of sending a QoS 1 message to its chosen subscribing Client and the connection to that Client breaks before the Server has received an acknowledgement from the Client, the Server MAY wait for the Client to reconnect and retransmit the message to that Client. If the Client'sSession terminates before the Client reconnects, the Server SHOULD send the Application Message to another Client that is subscribed to the same Shared Subscription. It MAY attempt to send the message to another Client as soon as it loses its connection to the first Client.

I can't be sure but I think that with current implementation VerneMQ will not pass a QoS 1 message to a different client in case of a missing PUBACK. Or I may be wrong and this will work but only for `clean=true` session. Still... even when `clean=false` such message could be sent to other online client and `at least once` would be still valid.

**Comment by mwilniewiec** [comment]
I recently tested different brokers performance & resource consumption. I have quite a good setup and I tested how HiveMQ-CE & EMQX are working in this area.
The results tells me that every broker acts differently.

1. Im starting 2 shared subscribers on QOS=1 and 1 publisher on QOS=1
2. I check if messages are distributed correctly   **(VerneMq, EMQX, HiveMQ - all OK)**
3. I close 1 subscriber
4. I check if all messages go to the online subscriber  **(VerneMq, EMQX (only with extra flag), HiveMQ - OK)**
5. I close second subscriber
6. I wait for ~10 messages to be published
7. I start only one subscriber
8. I check if subscriber gets all of the offline messages or just a part of it **(HiveMQ - All message as I would expect, VerneMQ - just a part, EMQX - no offline messages)**

In that case I would still say that implementation of HiveMQ is the only valid one that can be used for horizontally autoscaled & load-balanced subscribers.

@ioolkos What are your thoughts?

EDIT:
See, https://github.com/emqx/emqx/issues/2143#issuecomment-864878656
I see I'm not the only one :) and VerneMQ is not the only broker facing such problems :)

**Reply by ioolkos** [reply]
My thoughts are that VerneMQ has a valid implementation of Shared subscriptions. :)

[Why use shared subscriptions in the first place, when that single consumer coming back is able to consume the full message load? Shared subscriptions are consumer side load balancing, addressing the dreaded MQTT slow consumer problem. Maybe the use case of mixed scenarios (persistent/non-persistent) starts to make sense then, where you could keep the set of consumers you want to autoscale non-persistent while having the main set use persistency].

**Reply by ioolkos** [reply]
To add to this, @mwilniewiec please note that I value your testing & input, of course. Not dismissing it. 
But any alternative implementation of shared subscribers would likely require a new queue type. As such, an alternative implementation of shared subscribers is not on the roadmap.

**Reply by mwilniewiec** [reply]
I can't agree with that, your example is only theoretical and I think you misunderstood a slow consumer problem.
I'll provide a practical one...

You are building a "change your voice" application where people upload their voice in mp3 and can choose if they want to be a baby, granny, santa claus or squirrel. 
You implement a service which can modify those mp3s into a chosen voice.
It takes around 30 seconds to process the audio using your algorithm, http request-response is not an option.
You decide user action to publish an MQTT message to the broker with a link to uploaded mp3 + chosen voice info.
Your service subscribes to these messages, downloads mp3, processes it, uploads it somewhere and publishes a message that it's done attaching new link in a payload.
As you have only 10 users daily there is no problem. User waits for their modified mp3 at most 1 minute.
Your are getting more and more users and you see that during rush hours (7pm-10pm) people are waiting an hour for their mp3.
Your architecture is good so you can scale horizontally. You decide to put 3 workers, but ...  why would you keep 3 workers during the night when everybody is sleeping? You set autoscaler to 3 from 7pm-10pm but for the rest of the day 1 is enough.
Some users join, some users quit, different countries with different hours appear, situation is not that predictable any more.
You solve the problem by configuring scaling of your workers based on number of unprocessed messages. Lets say every 10 messages is 1 worker. During idle time just 1 is enough, during rush hours 12 are started as there are 120 mp3s to process concurrently.

And now we are getting to discussed problem where a power outage stops all your subscribers for 10 minutes.
In the meantime users were uploading their mp3s and there are 50 messages waiting to process.
Power is back, autoscaler starts 5 workers and they process 15 of those messages... What happened with the other 35? They are waiting for worker 6, 7, 8, 9, 10, 11, 12. Will they ever start? I may happen during the next rush hour but it is possible that it will never happen.

As my big arguments:
1. Lot of business use cases can be found with exactly the same problem. 
2. RabbitMQ, Kafka, HiveMQ can address this problem. You could say "Use RabbitMQ then", but MQTT has some other advantages and I can't see a reason why it should work differently in this area.

**Reply by ioolkos** [reply]
Happy to review any alternative implementation if proposed by PR. At least this would allow comparing the respective sets of pros/cons (aka tradeoffs).

> I think you misunderstood

No... 😎. I'm aware of the tradeoffs as stated.
